File: hardware_info.py
#!/usr/bin/env python
# -*-encoding:UTF8-*-
# Create by zhaozhang@yxlink.com
# On 2017/06/06

import sys
import base64
import logging
import MySQLdb
import ConfigParser
import os, fcntl, struct
from Crypto.Cipher import AES

logging.basicConfig(level=logging.ERROR,
                    format='%(asctime)s %(levelname)s %(filename)s[line:%(lineno)d] %(message)s')
WAF_CONFIG   = "/var/waf/waf.conf"
cfg    = ConfigParser.RawConfigParser()
cfg.readfp(open(WAF_CONFIG))
host   = cfg.get("mysql","db_ip").replace('"','')
user   = cfg.get("mysql","db_user").replace('"','')
passwd = cfg.get("mysql","db_passwd").replace('"','')

NET_70_FILE = "/etc/udev/rules.d/70-persistent-net.rules"

class PycryptoAES(object):
    _key = "vbglodckidapckcavbglodckidapckca"
    _iv = "eacpqkdckzefjalceacpqkdckzefjalc"

    def __init__(self,key=_key,iv=_iv[:16]):
        self.key = key
        self.iv = iv
        self.mode = AES.MODE_CBC
        self.BS = AES.block_size
        self.pad = lambda s: s + (self.BS - len(s) % self.BS)*chr(0)
        self.unpad = lambda s: s[0:-ord(s[-1])]

    def encrypt(self,text):
        try:
            text = self.pad(text)
            cryptor = AES.new(self.key,self.mode,self.iv)
            cipher_text = cryptor.encrypt(text)
            return base64.b64encode(cipher_text)
        except Exception,e:
            logging.DEBUG("failed encrypt():" + str(e))
            return ""

    def decrypt(self,text):
        try:
            cryptor = AES.new(self.key,self.mode,self.iv)
            plain_text = cryptor.decrypt(base64.b64decode(text))
            return plain_text.rstrip('\0')
        except Exception,e:
            logging.DEBUG("failed decrypt():" + str(e))
            return ""

class HardwareInfo(object):
    def __init__(self):
        self.core_num = self._cpu_core_number()
        # self.hdisk_sernum = self._hardisk_serial_num()
        self.cpu_type = self._cpu_type()
        self.cpu_id = self._cpu_id()
        self.mb_uuid = self._motherboard_uuid()
        self.sys_time = self._system_now()
        self.mem_total = self._memory_total()
        self.mac_list = self._ether_mac_list()

    def _hardisk_serial_num(self):
        try:
            fd = open("/dev/sda", "rb")
            hd_driveid_format_str = "@ 10H 20s 3H 8s 40s 2B H 2B H 4B 6H 2B I 36H I Q 152H"
            HDIO_GET_IDENTITY = 0x030d
            sizeof_hd_driveid = struct.calcsize(hd_driveid_format_str)
            assert sizeof_hd_driveid == 512
            buf = fcntl.ioctl(fd, HDIO_GET_IDENTITY, " " * sizeof_hd_driveid)
            fields = struct.unpack(hd_driveid_format_str, buf)
            serial_no = fields[10].strip()
            # model = fields[15].strip()
            fd.close()
            return serial_no
        except Exception,e:
            logging.DEBUG("get harddisk serial num failed:" + str(e))
            return ""

    def _cpu_core_number(self):
        try:
            ret = 0
            pip = os.popen("cat /proc/cpuinfo|grep pro|wc -l","r")
            if pip:
                ret = int(pip.read())
            pip.close()
            return ret
        except Exception,e:
            logging.DEBUG("get cpu core number failed," + str(e))
            return 0

    def _cpu_type(self):
        try:
            ret = ""
            pip = os.popen("cat /proc/cpuinfo|grep name|cut -f2 -d:|head -1","r")
            if pip:
                ret = str(pip.read()).strip()
            pip.close()
            return ret
        except Exception,e:
            logging.DEBUG("get cpu type failed," + str(e))
            return ""

    def _cpu_id(self):
        try:
            ret = ""
            pip = os.popen("dmidecode -t processor|grep ID|cut -d: -f2","r")
            if pip:
                ret = str(pip.read()).strip()
            pip.close()
            return ret
        except Exception,e:
            logging.DEBUG("get cpu id failed," + str(e))
            return ""

    def _motherboard_uuid(self):
        try:
            ret = ""
            pip = os.popen("dmidecode -s system-uuid","r")
            if pip:
                ret = str(pip.read()).strip()
            pip.close()
            return ret
        except Exception,e:
            logging.DEBUG("get mother board uuid failed," + str(e))
            return ""

    def _system_now(self):
        try:
            ret = ""
            pip = os.popen("date +%Y%m%d","r")
            if pip:
                ret = str(pip.read()).strip()
            pip.close()
            return ret
        except:
            logging.DEBUG("get system now failed," + str(e))
            return ""

    def _memory_total(self):
        try:
            ret = 0
            pip = os.popen("free -m|grep Mem|awk '{print int($2/1000+0.99)}'","r")
            if pip:
                ret = int(pip.read())
            pip.close()
            return ret
        except Exception,e:
            logging.DEBUG("get memory total failed," + str(e))
            return ""

    def _ether_mac_list(self):
        try:
            ret = ""
            if os.path.exists(NET_70_FILE):
                pip = os.popen("ifconfig -a|grep 'HWaddr'|grep -v wafbridge|awk '{print $5}'|sort","r")
                if pip:
                    for i in pip.readlines():
                        if not ret:
                            ret = i.rstrip("\n")
                        else:
                            ret += "," + i.rstrip("\n")
                pip.close()
            return ret
        except Exception,e:
            logging.DEBUG("get ether mac list failed," + str(e))
            return ""

def get_text_md5(text):
    try:
        import hashlib
        md = hashlib.md5()
        md.update(text)
        md5sum = md.hexdigest()
        return md5sum
    except Exception, e:
        logging.DEBUG("get_text_md5 failed" + str(e))
        return

def hardware_md5sum():
    try:
        tmp_dic = {}
        hardware = HardwareInfo()
        tmp_dic["core_num"] = hardware.core_num
        tmp_dic["mem_total"] = hardware.mem_total
        # tmp_dic["hdisk_sernum"] = hardware.hdisk_sernum
        tmp_dic["cpu_type"] = hardware.cpu_type
        tmp_dic["cpu_id"] = hardware.cpu_id
        tmp_dic["mb_uuid"] = hardware.mb_uuid
        tmp_dic["mac_list"] = hardware.mac_list

        dump_str = str(tmp_dic)
        md5_code = get_text_md5(dump_str)
        return md5_code
    except:
        return ""

def get_config_value(name):
    try:
        conn = MySQLdb.connect(host,user,passwd,db='waf_hw',charset='utf8')
        cur = conn.cursor(MySQLdb.cursors.DictCursor)
        sql = "select * from config where Name = '%s'" % name
        cur.execute(sql)
        ret = cur.fetchone()
        conn.close()

        if ret:
            return ret["Value"]
        else:
            return ""
    except Exception,e:
        logging.DEBUG("get_user_info error:" + str(e))
        return ""

def generator_challenge_code():
    try:
        info_dict = {}
        ret = ""
        hardware = HardwareInfo()
        info_dict["core_num"] = hardware.core_num
        info_dict["mem_total"] = hardware.mem_total
        info_dict["sys_time"] = hardware.sys_time
        # info_dict["hdisk_sernum"] = hardware.hdisk_sernum
        info_dict["cpu_type"] = hardware.cpu_type
        info_dict["cpu_id"] = hardware.cpu_id
        info_dict["mb_uuid"] = hardware.mb_uuid
        info_dict["mac_list"] = hardware.mac_list
        info_dict["user"] = get_config_value("user_name")
        info_dict["modelno"] = get_config_value("model_no")

        if info_dict:
            cryptor = PycryptoAES()
            dump_str = str(info_dict)
            ret = cryptor.encrypt(dump_str)
        return ret
    except:
        return ""

if __name__ == '__main__':

    if os.geteuid() >  0:
        # print("ERROR: Must be root to user")
        sys.exit(1)

    if len(sys.argv) == 2 and sys.argv[1] == "md5":
        md5_sum = hardware_md5sum()
        print md5_sum
    else:
        challenge_code = generator_challenge_code()
        print challenge_code

File: hardware_info.c
// @filename hardware_info.cpp
// Created by zhaozhang@yxlink.com
// on 2017/6/2.
//
#define _BSD_SOURCE	/* for strtoll() */
//#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
//#define __USE_GNU	/* for O_DIRECT */
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <linux/types.h>
#include <sys/ioctl.h>

#define	HDIO_GET_IDENTITY	 0x030d

static char *strip (char *s)
{
    char *e;

    while (*s == ' ') ++s;
    if (*s)
        for (e = s + strlen(s); *--e == ' '; *e = '\0');
    return s;
}

char *get_hard_disk_serno(char *devname)
{
    int fd;
    int err = 0;

    // char name[] = "/dev/sda";
    if (devname == NULL)
        return NULL;

    static int open_flags = O_RDONLY|O_NONBLOCK;

    fd = open(devname, open_flags);
    if (fd < 0)
    {
        err = errno;
        perror(devname);
    }
    else
    {
        __u16 id[256];

        if (!ioctl(fd, HDIO_GET_IDENTITY, id))
        {
            char *serno = strip(strndup((char *)&id[10], 20));

            //printf("HD serial: %s", serno);

            return serno;
        }
        else if (errno == -ENOMSG)
        {
            //printf("No identification info available\n");
        }
        else
        {
            err = errno;
            //perror("HDIO_GET_IDENTITY failed");
        }
    }
    return NULL;
}

#define cpuid(in,a,b,c,d) asm("cpuid": "=a" (a), "=b" (b), "=c" (c), "=d" (d) : "a" (in));
int get_cpu_id(char *cpu_id)
{
    unsigned long eax, ebx, ecx, edx;

    if(cpu_id == NULL){
        return -1;
    }
    cpuid (1, eax, ebx, ecx, edx);
    snprintf (cpu_id, 33, "%08lx%08lx%08lx%08lx", eax, ebx, ecx, edx);

    return 0;
}

int get_cpu_core_number()
{
    char buf[256];
    char cmd[] = "cat /proc/cpuinfo|grep pro|wc -l";
    FILE *fp = NULL;
    int ret = 0;

    if((fp = popen(cmd, "r")) != NULL) {
        if(fgets(buf, sizeof(buf), fp) != NULL){
            ret = atoi(buf);
        }
        pclose(fp);
    }

    return  ret;
}

int get_cpu_type(char *type)
{
    char buf[256];
    char cmd[] = "cat /proc/cpuinfo|grep name|cut -f2 -d:|head -1";
    FILE *fp = NULL;

    if(type == NULL){
        return -1;
    }

    if((fp = popen(cmd, "r")) != NULL) {
        if(fgets(buf, sizeof(buf), fp) != NULL){
            snprintf(type, sizeof(buf), "%s", buf);
        }
        pclose(fp);
    }

    return  0;
}

int get_motherboard_uuid(char *uuid)
{
    char buf[256];
    char cmd[] = "/usr/sbin/dmidecode -s system-uuid";
    FILE *fp = NULL;

    if(uuid == NULL){
        return -1;
    }

    if((fp = popen(cmd, "r")) != NULL) {
        if(fgets(buf, sizeof(buf), fp) != NULL){
            snprintf(uuid, sizeof(buf), "%s", buf);
        }
        pclose(fp);
    }

    return  0;
}

int get_mac_list(char *list)
{
    char buf[256];
    char cmd[] = "grep 'ATTR{address}' /etc/udev/rules.d/70-persistent-net.rules|awk -F ',' '{print $4}'|cut -d'\"' -f2";
    FILE *fp = NULL;
    char *tmp = list;

    if (list == NULL){
        return -1;
    }
    int idx = 0;
    if((fp = popen(cmd, "r")) != NULL) {
        while(fgets(buf, sizeof(buf), fp) != NULL){
            printf("mac is: %s", buf);
            int len = strlen(buf); // replace '\n' to ','
            buf[len-1] = ',';
            sprintf(tmp, "%s", buf);
            tmp = tmp + strlen(buf);
        }

        pclose(fp);
    }

    return  0;
}

int get_memory_total()
{
    char buf[256];
    char cmd[] = "free -m|grep Mem|awk '{print int($2/1000+0.99)}'";
    FILE *fp = NULL;
    int ret = 0;

    if((fp = popen(cmd, "r")) != NULL) {
        if(fgets(buf, sizeof(buf), fp) != NULL){
            ret = atoi(buf);
        }
        pclose(fp);
    }

    return  ret;
}

int get_system_time_now(char *time)
{
    char buf[256];
    char cmd[] = "/bin/date +%Y%m%d";
    FILE *fp = NULL;

    if(time == NULL){
        return -1;
    }

    if((fp = popen(cmd, "r")) != NULL) {
        if(fgets(buf, sizeof(buf), fp) != NULL){
            snprintf(time, sizeof(buf), "%s", buf);
        }
        pclose(fp);
    }

    return  0;
}

int main()
{
    char *serno = NULL;
    char hd_name[] = "/dev/sda";
    char cpuid[128];
    char motherboard_uuid[128];
    char mac_list[256];
    char cpu_type[256];
    char time[128];

    serno = get_hard_disk_serno(hd_name);
    printf("hard disk %s serno is: %s\n", hd_name, serno);

    get_cpu_id(cpuid);
    printf("cpu id is: %s\n", cpuid);

    get_motherboard_uuid(motherboard_uuid);
    printf("motherboard uuid is :%s\n", motherboard_uuid);

    get_mac_list(mac_list);
    printf("ehter mac list is :%s\n", mac_list);

    get_cpu_type(cpu_type);
    printf("cpu type is :%s\n", cpu_type);

    int num = get_cpu_core_number();
    printf("cpu num is : %d\n", num);

    int total = get_memory_total();
    printf("memory total is :%d\n", total);

    get_system_time_now(time);
    printf("system time now is: %s\n", time);

}

